File: agent/mcp/connection.py
from __future__ import annotations

import os
from abc import ABC, abstractmethod
from typing import Any

from loguru import logger

from .config import ServerConfig

try:
    from mcp import ClientSession, StdioServerParameters
    from mcp.client.stdio import stdio_client
    from mcp.client.sse import sse_client
    MCP_AVAILABLE = True
except ImportError:
    MCP_AVAILABLE = False
    ClientSession = None  # type: ignore[misc,assignment]
    StdioServerParameters = None  # type: ignore[misc,assignment]


class MCPConnection(ABC):
    """单个 MCP 服务器的连接抽象。"""

    def __init__(self, server_name: str, config: ServerConfig) -> None:
        self.server_name = server_name
        self.config = config
        self._session: Any = None
        self._client_ctx: Any = None
        self._connected = False

    @property
    def connected(self) -> bool:
        return self._connected

    @abstractmethod
    async def _enter_client(self) -> Any:
        """返回异步上下文管理器，其 __aenter__ 产生 (read_stream, write_stream)。"""
        ...

    async def connect(self) -> bool:
        if not MCP_AVAILABLE:
            logger.error(f"[MCP:{self.server_name}] mcp SDK not installed")
            return False
        try:
            self._client_ctx = await self._enter_client().__aenter__()
            read_stream, write_stream = self._client_ctx
            self._session = ClientSession(read_stream, write_stream)
            await self._session.__aenter__()
            await self._session.initialize()
            self._connected = True
            logger.info(f"[MCP:{self.server_name}] connected")
            return True
        except Exception as e:
            logger.warning(f"[MCP:{self.server_name}] connection failed: {e}")
            self._connected = False
            return False

    async def disconnect(self) -> None:
        if self._session is not None:
            try:
                await self._session.__aexit__(None, None, None)
            except Exception as e:
                logger.debug(f"[MCP:{self.server_name}] session cleanup error: {e}")
            self._session = None
        if self._client_ctx is not None:
            try:
                await self._client_ctx.__aexit__(None, None, None)
            except Exception as e:
                logger.debug(f"[MCP:{self.server_name}] client cleanup error: {e}")
            self._client_ctx = None
        self._connected = False
        logger.info(f"[MCP:{self.server_name}] disconnected")

    async def list_tools(self) -> list[dict[str, Any]]:
        if self._session is None or not self._connected:
            return []
        try:
            result = await self._session.list_tools()
            return [
                {
                    "name": t.name,
                    "description": t.description or "",
                    "inputSchema": t.inputSchema,
                }
                for t in result.tools
            ]
        except Exception as e:
            logger.warning(f"[MCP:{self.server_name}] list_tools failed: {e}")
            return []

    async def call_tool(self, tool_name: str, arguments: dict[str, Any]) -> str:
        if self._session is None or not self._connected:
            raise RuntimeError(f"MCP server '{self.server_name}' not connected")
        result = await self._session.call_tool(tool_name, arguments=arguments)
        texts: list[str] = []
        for item in result.content:
            if getattr(item, "type", None) == "text":
                texts.append(item.text)
            else:
                texts.append(str(item))
        output = "\n".join(texts) or "(empty result)"
        if result.isError:
            return f"Error: {output}"
        return output


# 允许传递给 MCP 子进程的环境变量白名单
_SAFE_ENV_KEYS = frozenset(
    {"PATH", "HOME", "USER", "SHELL", "LANG", "LC_ALL", "TMPDIR", "TERM", "PWD"}
)


class StdioConnection(MCPConnection):
    """通过子进程 stdio 连接 MCP 服务器。"""

    async def _enter_client(self) -> Any:
        base_env = {k: v for k, v in os.environ.items() if k in _SAFE_ENV_KEYS}
        if self.config.env:
            base_env.update(self.config.env)
        params = StdioServerParameters(
            command=self.config.command or "",
            args=list(self.config.args),
            env=base_env or None,
        )
        return stdio_client(params)


class SSEConnection(MCPConnection):
    """通过 SSE HTTP 连接 MCP 服务器。"""

    async def _enter_client(self) -> Any:
        return sse_client(
            url=self.config.url or "",
            headers=self.config.headers,
        )

File: packages/core/src/mcp/connection.ts
import type { ServerConfig } from './config'
import { Client } from '@modelcontextprotocol/sdk/client/index.js'
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js'
import { SSEClientTransport } from '@modelcontextprotocol/sdk/client/sse.js'
import type { ExecutionEnvironment } from '../environment/snapshot'

export interface MCPToolDefinition {
  name: string
  description?: string
  inputSchema: Record<string, unknown>
}

export interface MCPCallToolResult {
  content: string
  isError: boolean
}

export const SAFE_ENV_KEYS = new Set([
  'PATH',
  'HOME',
  'USER',
  'SHELL',
  'LANG',
  'LC_ALL',
  'TMPDIR',
  'TERM',
  'PWD',
  'USERPROFILE',
  'USERNAME',
  'SystemRoot',
  'ComSpec',
  'PATHEXT',
  'TEMP',
  'TMP',
])

export abstract class MCPConnection {
  readonly serverName: string
  connected = false
  private activeCalls = 0
  private environmentRevision: string | null = null
  private environmentQueue: Promise<void> = Promise.resolve()
  private readonly idleWaiters = new Set<() => void>()

  constructor(serverName: string) {
    this.serverName = serverName
  }

  abstract connect(): Promise<boolean>
  abstract disconnect(): Promise<void>
  abstract listTools(): Promise<MCPToolDefinition[]>
  abstract callTool(
    toolName: string,
    args: Record<string, unknown>,
  ): Promise<MCPCallToolResult>

  get executionEnvironmentRevision(): string | null {
    return this.environmentRevision
  }

  async callToolWithEnvironment(
    toolName: string,
    args: Record<string, unknown>,
    snapshot: ExecutionEnvironment,
  ): Promise<MCPCallToolResult> {
    await this.prepareExecutionEnvironment(snapshot)
    this.activeCalls += 1
    try {
      return await this.callTool(toolName, args)
    } finally {
      this.activeCalls -= 1
      if (this.activeCalls === 0) {
        for (const resolve of this.idleWaiters) resolve()
        this.idleWaiters.clear()
      }
    }
  }

  protected async applyExecutionEnvironment(
    _snapshot: ExecutionEnvironment,
  ): Promise<void> {}

  protected adoptExecutionEnvironment(snapshot: ExecutionEnvironment): void {
    this.environmentRevision = snapshot.revision
  }

  private async prepareExecutionEnvironment(
    snapshot: ExecutionEnvironment,
  ): Promise<void> {
    const operation = this.environmentQueue.then(async () => {
      if (this.environmentRevision === snapshot.revision) return
      if (this.activeCalls > 0)
        await new Promise<void>((resolve) => this.idleWaiters.add(resolve))
      if (this.environmentRevision === snapshot.revision) return
      await this.applyExecutionEnvironment(snapshot)
      this.environmentRevision = snapshot.revision
    })
    this.environmentQueue = operation.catch(() => {})
    await operation
  }
}

export function buildStdioEnv(
  config: Pick<ServerConfig, 'env'> | { env?: Record<string, string> },
  env: Record<string, string | undefined> = process.env,
): Record<string, string> {
  const out: Record<string, string> = {}
  for (const [key, value] of Object.entries(env)) {
    if (value !== undefined && SAFE_ENV_KEYS.has(key)) out[key] = value
  }
  for (const [key, value] of Object.entries(config.env ?? {})) out[key] = value
  return out
}

export class StdioConnection extends MCPConnection {
  config: ServerConfig
  private client: Client | null = null
  private executionEnvironment: ExecutionEnvironment | null
  private readonly configResolver:
    ((snapshot: ExecutionEnvironment) => ServerConfig | null) | null

  constructor(
    serverName: string,
    config: ServerConfig,
    opts: {
      executionEnvironment?: ExecutionEnvironment | null
      configResolver?:
        ((snapshot: ExecutionEnvironment) => ServerConfig | null) | null
    } = {},
  ) {
    super(serverName)
    this.config = config
    this.executionEnvironment = opts.executionEnvironment ?? null
    this.configResolver = opts.configResolver ?? null
    if (this.executionEnvironment)
      this.adoptExecutionEnvironment(this.executionEnvironment)
  }

  stdioParams(env: Record<string, string | undefined> = process.env): {
    command: string
    args: string[]
    env: Record<string, string> | undefined
  } {
    const childEnv = buildStdioEnv(this.config, env)
    return {
      command: this.config.command ?? '',
      args: this.config.args,
      env: Object.keys(childEnv).length ? childEnv : undefined,
    }
  }

  async connect(): Promise<boolean> {
    try {
      const transport = new StdioClientTransport({
        ...this.stdioParams(this.executionEnvironment?.env ?? process.env),
        stderr: 'inherit',
      })
      const client = new Client({ name: 'emperor-agent', version: '0.0.0' })
      await client.connect(transport)
      this.client = client
      this.connected = true
      return true
    } catch {
      this.client = null
      this.connected = false
      return false
    }
  }

  async disconnect(): Promise<void> {
    await this.client?.close().catch(() => {})
    this.client = null
    this.connected = false
  }

  async listTools(): Promise<MCPToolDefinition[]> {
    if (!this.client || !this.connected) return []
    try {
      const result = await this.client.listTools()
      return result.tools.map((tool) => ({
        name: tool.name,
        description: tool.description ?? '',
        inputSchema: tool.inputSchema as Record<string, unknown>,
      }))
    } catch {
      return []
    }
  }

  async callTool(
    toolName: string,
    args: Record<string, unknown>,
  ): Promise<MCPCallToolResult> {
    if ((!this.client || !this.connected) && !(await this.connect()))
      throw new Error(`MCP server '${this.serverName}' not connected`)
    const client = this.client
    if (!client)
      throw new Error(`MCP server '${this.serverName}' not connected`)
    return normalizeCallToolResult(
      await client.callTool({ name: toolName, arguments: args }),
    )
  }

  protected override async applyExecutionEnvironment(
    snapshot: ExecutionEnvironment,
  ): Promise<void> {
    const resolvedConfig = this.configResolver?.(snapshot)
    if (this.configResolver && !resolvedConfig)
      throw new Error(`MCP server '${this.serverName}' is no longer configured`)
    if (resolvedConfig) this.config = resolvedConfig
    const reconnect = this.connected
    this.executionEnvironment = snapshot
    if (!reconnect) return
    await this.disconnect()
    if (!(await this.connect()))
      throw new Error(`MCP server '${this.serverName}' failed to reconnect`)
  }
}

export class SSEConnection extends MCPConnection {
  readonly config: ServerConfig
  private client: Client | null = null

  constructor(serverName: string, config: ServerConfig) {
    super(serverName)
    this.config = config
  }

  async connect(): Promise<boolean> {
    try {
      if (!this.config.url) throw new Error('missing MCP SSE url')
      const transport = new SSEClientTransport(new URL(this.config.url), {
        eventSourceInit:
          this.config.headers && Object.keys(this.config.headers).length
            ? ({ fetch: withHeaders(this.config.headers) } as never)
            : undefined,
        requestInit: Object.keys(this.config.headers).length
          ? { headers: this.config.headers }
          : undefined,
      })
      const client = new Client({ name: 'emperor-agent', version: '0.0.0' })
      await client.connect(transport)
      this.client = client
      this.connected = true
      return true
    } catch {
      this.client = null
      this.connected = false
      return false
    }
  }

  async disconnect(): Promise<void> {
    await this.client?.close().catch(() => {})
    this.client = null
    this.connected = false
  }

  async listTools(): Promise<MCPToolDefinition[]> {
    if (!this.client || !this.connected) return []
    try {
      const result = await this.client.listTools()
      return result.tools.map((tool) => ({
        name: tool.name,
        description: tool.description ?? '',
        inputSchema: tool.inputSchema as Record<string, unknown>,
      }))
    } catch {
      return []
    }
  }

  async callTool(
    toolName: string,
    args: Record<string, unknown>,
  ): Promise<MCPCallToolResult> {
    if (!this.client || !this.connected)
      throw new Error(`MCP server '${this.serverName}' not connected`)
    return normalizeCallToolResult(
      await this.client.callTool({ name: toolName, arguments: args }),
    )
  }
}

function normalizeCallToolResult(
  result: Awaited<ReturnType<Client['callTool']>>,
): MCPCallToolResult {
  if ('toolResult' in result)
    return {
      content: stringifyUnknown(result.toolResult),
      isError: Boolean(result.isError),
    }
  const content =
    'content' in result && Array.isArray(result.content) ? result.content : []
  const texts = content.map((item) => {
    if (item.type === 'text') return item.text
    if (item.type === 'resource' && 'resource' in item)
      return stringifyUnknown(item.resource)
    return stringifyUnknown(item)
  })
  if (
    !texts.length &&
    'structuredContent' in result &&
    result.structuredContent
  )
    texts.push(stringifyUnknown(result.structuredContent))
  const output = texts.join('\n') || '(empty result)'
  return { content: output, isError: Boolean(result.isError) }
}

function stringifyUnknown(value: unknown): string {
  return typeof value === 'string' ? value : JSON.stringify(value)
}

function withHeaders(headers: Record<string, string>): typeof fetch {
  return ((input: RequestInfo | URL, init?: RequestInit) => {
    const existing =
      init?.headers instanceof Headers
        ? Object.fromEntries(init.headers.entries())
        : init?.headers &&
            typeof init.headers === 'object' &&
            !Array.isArray(init.headers)
          ? (init.headers as Record<string, string>)
          : {}
    return fetch(input, { ...init, headers: { ...existing, ...headers } })
  }) as typeof fetch
}
